lib/collection/src/grouping/aggregator.rs

```diff
@@ -0,0 +1,380 @@
+use std::collections::{HashMap, HashSet};
+use std::hash::Hash;
+
+use segment::types::{ExtendedPointId, ScoredPoint};
+use serde_json::Value;
+use AggregatorError::*;
+
+pub(super) enum AggregatorError {
+    GroupsFull,
+    Other,
+}
+
+/// Abstraction over serde_json::Value to be used as a key in a HashMap/HashSet
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub(super) struct GroupKey(pub serde_json::Value);
+
+impl TryFrom<serde_json::Value> for GroupKey {
+    type Error = AggregatorError;
+
+    /// Only allows Strings and Numbers to be converted into GroupKey
+    fn try_from(value: serde_json::Value) -> Result<Self, Self::Error> {
+        match value {
+            serde_json::Value::String(_) | serde_json::Value::Number(_) => Ok(Self(value)),
+            _ => Err(Other),
+        }
+    }
+}
+
+impl Hash for GroupKey {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        match &self.0 {
+            Value::Number(n) => n.hash(state),
+            Value::String(s) => s.hash(state),
+            _ => unreachable!("GroupKey should only be a number or a string"),
+        }
+    }
+}
+
+type Hits = HashSet<ScoredPoint>;
+
+#[allow(dead_code)] // temporary
+pub(super) struct GroupsAggregator {
+    groups: HashMap<GroupKey, Hits>,
+    max_group_size: usize,
+    grouped_by: String,
+    max_groups: usize,
+    full_groups: HashSet<GroupKey>,
+}
+
+#[allow(dead_code)] // temporary
+impl GroupsAggregator {
+    pub(super) fn new(groups: usize, group_size: usize, grouped_by: String) -> Self {
+        Self {
+            groups: HashMap::with_capacity(groups),
+            max_group_size: group_size,
+            grouped_by,
+            max_groups: groups,
+            full_groups: HashSet::with_capacity(groups),
+        }
+    }
+
+    /// Adds a point to the group that corresponds based on the group_by field, assumes that the point has the group_by field
+    fn add_point(&mut self, point: &ScoredPoint) -> Result<(), AggregatorError> {
+        if self.full_groups.len() == self.max_groups {
+            return Err(GroupsFull);
+        }
+
+        // if the key contains multiple values, grabs the first one
+        let group_key = point
+            .payload
+            .as_ref()
+            .and_then(|p| p.get_value(&self.grouped_by).next().cloned())
+            .ok_or(Other)
```

Comment:
`Other` would deserve to be more concrete in that case, something like `KeyNotFound` maybe.